File: scripts/check-links.py
#!/usr/bin/env python3
"""
Simple link checker using the reliable linkchecker package.
Much simpler and more reliable than custom Node.js solutions.
"""

import subprocess
import sys
import os
import argparse
from pathlib import Path

def run_command(cmd, description):
    """Run a command and handle errors gracefully."""
    print(f"🔍 {description}...")
    try:
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=300)
        if result.returncode == 0:
            print(f"✅ {description} - OK")
            return True
        else:
            print(f"❌ {description} - FAILED")
            if result.stderr:
                print(f"Error: {result.stderr}")
            if result.stdout:
                print(f"Output: {result.stdout}")
            return False
    except subprocess.TimeoutExpired:
        print(f"⏰ {description} - TIMEOUT")
        return False
    except Exception as e:
        print(f"💥 {description} - ERROR: {e}")
        return False

def check_prerequisites():
    """Check if required tools are available."""
    print("🔧 Checking prerequisites...")
    
    # Check Python 3
    if not run_command("python3 --version", "Python 3 availability"):
        sys.exit(1)
    
    # Check linkchecker
    if not run_command("linkchecker --version", "linkchecker availability"):
        print("📦 Installing linkchecker...")
        if not run_command("pip3 install linkchecker", "linkchecker installation"):
            print("❌ Failed to install linkchecker. Please install manually:")
            print("   pip3 install linkchecker")
            sys.exit(1)

def start_server():
    """Start a local HTTP server."""
    print("🚀 Starting local server on port 8080...")
    try:
        # Check if server is already running
        result = subprocess.run("curl -s http://localhost:8080 > /dev/null", shell=True)
        if result.returncode == 0:
            print("✅ Server already running on port 8080")
            return None
        
        # Start new server
        server = subprocess.Popen(
            ["python3", "-m", "http.server", "8080"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL
        )
        
        # Give server time to start
        import time
        time.sleep(3)
        
        # Verify server is running
        if subprocess.run("curl -s http://localhost:8080 > /dev/null", shell=True).returncode == 0:
            print("✅ Server started successfully")
            return server
        else:
            print("❌ Failed to start server")
            server.terminate()
            sys.exit(1)
            
    except Exception as e:
        print(f"❌ Server startup error: {e}")
        sys.exit(1)

def check_links(internal_only=False, verbose=False):
    """Check links using linkchecker."""
    base_url = "http://localhost:8080"
    
    # Find HTML files to check
    html_files = []
    for file_path in ["index.html"]:
        if os.path.exists(file_path):
            html_files.append(f"{base_url}/{file_path}")
    
    # Add subdirectory HTML files
    for subdir in ["book", "tinytorch", "labs", "kits"]:
        html_file = Path(subdir) / "index.html"
        if html_file.exists():
            html_files.append(f"{base_url}/{subdir}/index.html")
    
    if not html_files:
        print("❌ No HTML files found to check")
        return False, False
    
    internal_success = True
    external_success = True
    
    # Check internal links first
    for url in html_files:
        print(f"\n🔗 Checking internal links for {url}...")
        
        cmd_parts = [
            "linkchecker",
            "--no-warnings",
            "--output=text",
            url
        ]
        
        if verbose:
            print(f"Running: {' '.join(cmd_parts)}")
        
        try:
            result = subprocess.run(
                cmd_parts,
                capture_output=True,
                text=True,
                timeout=120
            )
            
            if result.returncode == 0:
                print(f"✅ {url} - Internal links OK")
            else:
                print(f"❌ {url} - Found broken internal links")
                if result.stdout.strip():
                    print(result.stdout)
                if result.stderr.strip():
                    print("STDERR:", result.stderr)
                internal_success = False
                
        except subprocess.TimeoutExpired:
            print(f"⏰ {url} - Internal link check timed out")
            internal_success = False
        except Exception as e:
            print(f"💥 {url} - Internal link check error: {e}")
            internal_success = False
    
    # Check external links separately (non-blocking)
    if not internal_only:
        print(f"\n🌐 Checking external links (non-blocking)...")
        for url in html_files:
            cmd_parts = [
                "linkchecker",
                "--check-extern",
                "--timeout=10",
                "--threads=3",
                "--no-warnings",
                "--output=text",
                url
            ]
            
            if verbose:
                print(f"Running: {' '.join(cmd_parts)}")
            
            try:
                result = subprocess.run(
                    cmd_parts,
                    capture_output=True,
                    text=True,
                    timeout=60  # Shorter timeout for external
                )
                
                if result.returncode == 0:
                    print(f"✅ {url} - External links OK")
                else:
                    print(f"⚠️  {url} - Some external links may be broken (non-critical)")
                    if verbose and result.stdout.strip():
                        print(result.stdout)
                    external_success = False
                    
            except subprocess.TimeoutExpired:
                print(f"⏰ {url} - External link check timed out (non-critical)")
                external_success = False
            except Exception as e:
                print(f"⚠️  {url} - External link check error (non-critical): {e}")
                external_success = False
    
    return internal_success, external_success

def check_common_issues():
    """Check for common link problems."""
    print("\n🔎 Checking for common issues...")
    
    html_files = ["index.html"]
    for subdir in ["book", "tinytorch", "labs", "kits"]:
        html_file = Path(subdir) / "index.html"
        if html_file.exists():
            html_files.append(str(html_file))
    
    issues_found = False
    
    for file_path in html_files:
        if not os.path.exists(file_path):
            continue
            
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Check for doubled protocols
        if 'hhttps' in content or 'hhttp' in content:
            print(f"❌ {file_path}: Found doubled protocol (hhttps/hhttp)")
            issues_found = True
        
        # Check for localhost links
        if 'localhost' in content or '127.0.0.1' in content:
            print(f"❌ {file_path}: Found localhost links")
            issues_found = True
    
    if not issues_found:
        print("✅ No common issues found")
    
    return not issues_found

def main():
    parser = argparse.ArgumentParser(description="Check website links")
    parser.add_argument("--internal-only", action="store_true", 
                       help="Check only internal links (faster)")
    parser.add_argument("--verbose", action="store_true",
                       help="Verbose output")
    args = parser.parse_args()
    
    print("🔗 MLSysBook Link Checker (Python Edition)")
    print("=" * 45)
    
    if args.internal_only:
        print("ℹ️  Internal links only (faster mode)")
    
    try:
        # Check prerequisites
        check_prerequisites()
        
        # Start server
        server = start_server()
        
        try:
            # Check common issues
            common_ok = check_common_issues()
            
            # Check links
            internal_ok, external_ok = check_links(args.internal_only, args.verbose)
            
            # Final result
            if common_ok and internal_ok:
                if not external_ok and not args.internal_only:
                    print("\n⚠️  Some external links may be broken, but that's okay!")
                    print("Internal links are all working - safe to deploy.")
                else:
                    print("\n🎉 All link checks passed!")
                    print("Your website is ready for deployment.")
                sys.exit(0)
            else:
                print("\n❌ Critical internal link issues found!")
                print("Please fix internal links and common issues before deploying.")
                print("(External link issues are non-critical and won't block deployment)")
                sys.exit(1)
                
        finally:
            # Clean up server
            if server:
                print("\n🧹 Stopping server...")
                server.terminate()
                server.wait()
                
    except KeyboardInterrupt:
        print("\n👋 Link check interrupted by user")
        sys.exit(0)

if __name__ == "__main__":
    main() 

File: scripts/check-links.js
#!/usr/bin/env node

const { execSync, spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

// Configuration
const BASE_PORT = 8080;
const TIMEOUT = 30000; // 30 seconds
const args = process.argv.slice(2);
const internalOnly = args.includes('--internal-only');
const verbose = args.includes('--verbose');

// Colors for output
const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m',
  bold: '\x1b[1m'
};

function log(message, color = 'reset') {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

function checkPrerequisites() {
  try {
    execSync('which python3', { stdio: 'ignore' });
  } catch (error) {
    log('❌ Python3 is required but not found. Please install Python3.', 'red');
    process.exit(1);
  }

  // Check for broken-link-checker
  try {
    execSync('which blc', { stdio: 'ignore' });
  } catch (error) {
    // Try to find it in npm global location
    try {
      execSync('npm list -g broken-link-checker', { stdio: 'ignore' });
    } catch (npmError) {
      log('📦 Installing broken-link-checker globally...', 'yellow');
      try {
        execSync('npm install -g broken-link-checker', { stdio: 'inherit' });
        log('✅ broken-link-checker installed successfully!', 'green');
      } catch (installError) {
        log('❌ Failed to install broken-link-checker. Please run: npm install -g broken-link-checker', 'red');
        log('  Or ensure your npm global bin directory is in your PATH', 'yellow');
        process.exit(1);
      }
    }
  }
}

function startServer() {
  return new Promise((resolve, reject) => {
    log(`🚀 Starting local server on port ${BASE_PORT}...`, 'blue');
    
    // Check if port is already in use
    const testConnection = spawn('curl', ['-s', `http://localhost:${BASE_PORT}`], {
      stdio: 'ignore'
    });
    
    testConnection.on('close', (code) => {
      if (code === 0) {
        log(`✅ Server already running on port ${BASE_PORT}`, 'green');
        resolve({ kill: () => {} }); // Return dummy server object
        return;
      }
      
      // Start new server
      const server = spawn('python3', ['-m', 'http.server', BASE_PORT.toString()], {
        stdio: verbose ? 'inherit' : 'ignore'
      });

      // Give server time to start
      setTimeout(() => {
        resolve(server);
      }, 2000); // Reduced from 3000ms for faster pre-commit

      server.on('error', (error) => {
        log(`❌ Failed to start server: ${error.message}`, 'red');
        reject(error);
      });
    });
  });
}

function checkInternalLinks() {
  return new Promise((resolve, reject) => {
    log('🔍 Checking internal links...', 'blue');
    
    const baseUrl = `http://localhost:${BASE_PORT}`;
    const files = ['index.html'];
    
    // Add subdirectory HTML files
    const subdirs = ['book', 'tinytorch', 'labs', 'kits'];
    subdirs.forEach(dir => {
      const htmlFile = path.join(dir, 'index.html');
      if (fs.existsSync(htmlFile)) {
        files.push(htmlFile);
      }
    });

    let hasErrors = false;
    let completed = 0;
    const total = files.length;

    files.forEach(file => {
      const url = `${baseUrl}/${file}`;
      log(`  Checking ${file}...`, 'blue');
      
      const blc = spawn('blc', [
        url,
        '--recursive',
        '--exclude-external',
        '--filter-level', '3',
        '--requests', '2'
      ]);

      let output = '';
      blc.stdout.on('data', (data) => {
        output += data.toString();
      });

      blc.stderr.on('data', (data) => {
        output += data.toString();
      });

      blc.on('close', (code) => {
        completed++;
        
        if (code !== 0) {
          log(`    ❌ ${file}: Found broken internal links`, 'red');
          if (verbose) {
            console.log(output);
          }
          hasErrors = true;
        } else {
          log(`    ✅ ${file}: All internal links OK`, 'green');
        }

        if (completed === total) {
          if (hasErrors) {
            reject(new Error('Internal link check failed'));
          } else {
            resolve();
          }
        }
      });
    });
  });
}

function checkExternalLinks() {
  return new Promise((resolve, reject) => {
    log('🌐 Checking external links (this may take a while)...', 'blue');
    
    const url = `http://localhost:${BASE_PORT}/index.html`;
    const blc = spawn('blc', [
      url,
      '--external-only',
      '--filter-level', '2',
      '--requests', '1',
      '--exclude', 'linkedin.com',
      '--exclude', 'twitter.com',
      '--exclude', 'facebook.com'
    ]);

    let output = '';
    blc.stdout.on('data', (data) => {
      output += data.toString();
    });

    blc.stderr.on('data', (data) => {
      output += data.toString();
    });

    blc.on('close', (code) => {
      if (code !== 0) {
        log('⚠️  Some external links may be broken (non-critical)', 'yellow');
        if (verbose) {
          console.log(output);
        }
      } else {
        log('✅ All external links OK', 'green');
      }
      // Don't fail on external link issues
      resolve();
    });
  });
}

function checkCommonIssues() {
  log('🔎 Checking for common link issues...', 'blue');
  
  const files = ['index.html'];
  const subdirs = ['book', 'tinytorch', 'labs', 'kits'];
  subdirs.forEach(dir => {
    const htmlFile = path.join(dir, 'index.html');
    if (fs.existsSync(htmlFile)) {
      files.push(htmlFile);
    }
  });

  let hasIssues = false;

  files.forEach(file => {
    if (!fs.existsSync(file)) return;
    
    const content = fs.readFileSync(file, 'utf8');
    
    // Check for doubled protocols
    if (content.includes('hhttps') || content.includes('hhttp')) {
      log(`❌ ${file}: Found doubled protocol (hhttps/hhttp)`, 'red');
      hasIssues = true;
    }
    
    // Check for localhost links
    if (content.includes('localhost') || content.includes('127.0.0.1')) {
      log(`❌ ${file}: Found localhost links`, 'red');
      hasIssues = true;
    }
    
    // Check for unclosed href attributes (basic check)
    const unclosedHref = /href="[^"]*$/gm;
    if (unclosedHref.test(content)) {
      log(`❌ ${file}: Found unclosed href attributes`, 'red');
      hasIssues = true;
    }
  });

  if (!hasIssues) {
    log('✅ No common link issues found', 'green');
  }

  return !hasIssues;
}

async function main() {
  log('🔗 MLSysBook Link Checker', 'bold');
  log('=' .repeat(40), 'blue');
  
  if (internalOnly) {
    log('ℹ️  Running in internal-only mode (faster)', 'yellow');
  }

  try {
    // Check prerequisites
    checkPrerequisites();
    
    // Start local server
    const server = await startServer();
    
    try {
      // Check common issues first (fast)
      const commonIssuesOk = checkCommonIssues();
      
      // Check internal links
      await checkInternalLinks();
      
      // Check external links (unless internal-only mode)
      if (!internalOnly) {
        await checkExternalLinks();
      }
      
      // Final result
      if (commonIssuesOk) {
        log('\n🎉 All link checks passed!', 'green');
        log('Your website is ready for deployment.', 'green');
      } else {
        throw new Error('Common issues found');
      }
      
    } finally {
      // Clean up server
      log('🧹 Cleaning up server...', 'blue');
      server.kill();
    }
    
  } catch (error) {
    log(`\n❌ Link check failed: ${error.message}`, 'red');
    log('Please fix the issues above before deploying.', 'red');
    process.exit(1);
  }
}

// Handle Ctrl+C gracefully
process.on('SIGINT', () => {
  log('\n👋 Link check interrupted by user', 'yellow');
  process.exit(0);
});

main(); 